models/project_repository.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/jinzhu/gorm"
+	satoriuuid "github.com/satori/go.uuid"
+)
+
+// NewProjectRepository creates a new project repo
+func NewProjectRepository(db *gorm.DB) *GormProjectRepository {
+	return &GormProjectRepository{db}
+}
+
+// GormProjectRepository implements ProjectRepository using gorm
+type GormProjectRepository struct {
+	db *gorm.DB
+}
+
+// Load returns the project for the given id
+// returns NotFoundError or InternalError
+func (r *GormProjectRepository) Load(ctx context.Context, ID satoriuuid.UUID) (*project.Project, error) {
+	log.Printf("loading project %v", ID)
+	res := project.Project{}
+	tx := r.db.Where("id=?", ID).First(&res)
+	if tx.RecordNotFound() {
+		log.Printf("not found, res=%v", res)
+		return nil, NewNotFoundError("project", ID.String())
+	}
+	if tx.Error != nil {
+		return nil, NewInternalError(tx.Error.Error())
+	}
+	return &res, nil
+}
+
+// Delete deletes the project with the given id
+// returns NotFoundError or InternalError
+func (r *GormProjectRepository) Delete(ctx context.Context, ID satoriuuid.UUID) error {
+	project := project.Project{ID: ID}
+	tx := r.db.Delete(project)
+
+	if err := tx.Error; err != nil {
+		return NewInternalError(err.Error())
+	}
+	if tx.RowsAffected == 0 {
+		return NewNotFoundError("project", ID.String())
+	}
+
+	return nil
+}
+
+// Save updates the given project in the db. Version must be the same as the one in the stored version
+// returns NotFoundError, BadParameterError, VersionConflictError or InternalError
+func (r *GormProjectRepository) Save(ctx context.Context, p project.Project) (*project.Project, error) {
+	pr := project.Project{}
+	tx := r.db.Where("id=?", p.ID).First(&pr)
+	oldVersion := p.Version
+	p.Version++
+	if tx.RecordNotFound() {
+		// treating this as a not found error: the fact that we're using number internal is implementation detail
+		return nil, NotFoundError{"project", p.ID.String()}
+	}
+	if err := tx.Error; err != nil {
+		log.Print(err.Error())
+		return nil, InternalError{simpleError{err.Error()}}
+	}
+	tx = tx.Where("Version = ?", oldVersion).Save(&p)
+	if err := tx.Error; err != nil {
+		if isCheckViolation(tx.Error, "projects_name_check") {
+			return nil, NewBadParameterError("Name", p.Name).Expected("not empty")
+		}
+		if isUniqueViolation(tx.Error, "projects_name_idx") {
+			return nil, NewBadParameterError("Name", p.Name).Expected("unique")
+		}
+		return nil, InternalError{simpleError{err.Error()}}
+	}
+	if tx.RowsAffected == 0 {
+		return nil, VersionConflictError{simpleError{"version conflict"}}
+	}
+	log.Printf("updated project to %v\n", p)
+	return &p, nil
+}
+
+// Create creates a new Project in the db
+// returns BadParameterError or InternalError
+func (r *GormProjectRepository) Create(ctx context.Context, name string) (*project.Project, error) {
+	newProject := project.Project{
+		Name: name,
+	}
+
+	tx := r.db.Create(&newProject)
+	if err := tx.Error; err != nil {
+		log.Print(err.Error())
```

Comment:
That is true for Java, where you usually include the original exception (including the throw location). With the way our errors are constructed now, we might lose info when transforming errors.
However, I agree as it is here, the "log" call is pretty useless.